File: 127-word-ladder/word-ladder.py
import queue
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        q = queue.Queue()
        store = set(wordList)
        q.put((beginWord, 1))
        count = 0
        while(not q.empty()):
            curr = q.get()
            if curr[0] == endWord:
                return curr[1]
            word = list(curr[0])
            count = curr[1]
            for i in range(len(word)):
                l = word[i]
                for ch in string.ascii_lowercase:
                    word[i] = ch
                    result = ''.join(word)
                    if result in store:
                        q.put((result, count+1))
                        store.remove(result)
                word[i] = l
        return 0

File: 127-word-ladder/word-ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        queue<pair<string,int>> q;
        set<string> st(wordList.begin(), wordList.end());
        q.push({beginWord, 1});
        while(!q.empty()){
            string curr = q.front().first;
            int count = q.front().second;
            string copy = curr;
            if (curr == endWord){
                return count;
            }
            q.pop();
            for(int index = 0; index < curr.size(); index++){
                for(char c = 'a'; c <='z'; c++){
                    copy[index] = c;
                    if(st.find(copy) != st.end()){
                        q.push({copy, count+ 1});
                        st.erase(copy);
                    }
                    copy = curr;
                }
            }
        }
        return 0;
    }
};